src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_star_simplex_iterators.h

```diff
@@ -0,0 +1,285 @@
+  bool equal(Simplex_tree_optimized_cofaces_rooted_subtrees_simplex_iterator const& other) const {
+    if (other.st_ == nullptr) {
+      return (st_ == nullptr);
+    }
+    if (st_ == nullptr) {
+      return false;
+    }
+    return (it_ == other.it_);
+  }
+
+  Simplex_handle const& dereference() const { return sh_; }
+
+  void increment() {
+    if (++it_ == end_) {
+      st_ = nullptr;
+    }       //== end
+    else {  // update sh_
+      Node& curr_node = static_cast<Node&>(*it_);
+      auto curr_sib = st_->self_siblings(st_->simplex_handle_from_node(curr_node));
+      sh_ = curr_sib->find(max_v_);
+    }
+  }
+
+  // given a Node of label max_v, returns true if the associated simplex is a coface of the simplex {..., max_v}. The
+  // predicate stores the vertices of the simplex whose star we compute.
+  is_coface predicate_;
+  SimplexTree* st_;
+  // filtered iterators over Nodes of same label max_v_, filtered with predicate_
+  Filtered_cofaces_simplex_iterator it_;
+  Filtered_cofaces_simplex_iterator end_;
+  // max label of the simplex whose cofaces are computed
+  Vertex_handle max_v_;
+  // current Simplex_handle corresponding to Node pointed at by it_
+  Simplex_handle sh_;
+};
+
+/** \brief Iterator over the simplices of the star of a simplex.
+ *
+ * \details All Nodes of the simplex must store intrusive list hooks (option
+ * link_nodes_by_label == true) in order to connect all nodes with a fixed label u
+ * into a list.
+ *
+ * Forward iterator, value_type is SimplexTree::Simplex_handle.
+ *
+ * The implementation uses a
+ * Simplex_tree_optimized_cofaces_rooted_subtrees_simplex_iterator to iterate through all
+ * roots of cofaces Nodes, and traverses each such subtree of the simplex tree.
+ */
+template <class SimplexTree>
+class Simplex_tree_optimized_star_simplex_iterator
+    : public boost::iterator_facade<Simplex_tree_optimized_star_simplex_iterator<SimplexTree>,
+                                    typename SimplexTree::Simplex_handle const, boost::forward_traversal_tag> {
+ public:
+  using Simplex_handle = typename SimplexTree::Simplex_handle;
+  using Siblings = typename SimplexTree::Siblings;
+  using Vertex_handle = typename SimplexTree::Vertex_handle;
+  using Node = typename SimplexTree::Node;
+  using Static_vertex_vector = typename SimplexTree::Static_vertex_vector;
+
+  // any end() iterator
+  Simplex_tree_optimized_star_simplex_iterator() : st_(nullptr) {}
+
+  Simplex_tree_optimized_star_simplex_iterator(SimplexTree* cpx, Static_vertex_vector&& simp)
+      : st_(cpx), it_(cpx, std::move(simp)), end_(), sh_(*it_), sib_(st_->self_siblings(sh_)), bfs_queue_() {
+    if (it_ == end_) {
+      st_ = nullptr;
+      return;
+    }  // no coface subtree => end()
+    is_root_ = true;
+    sh_ = *it_;                      // sh_ is the root
+    sib_ = st_->self_siblings(sh_);  // Siblings containing sh_
+    if (st_->has_children(sh_)) {
+      bfs_queue_.insert(bfs_queue_.begin(), st_->children(sh_));
```

Comment:
You are trying to keep it as a queue, but I am not convinced that the order actually matters. Assuming it doesn't, you can handle it as a stack and just push_back when you want to add an element.

Reply:
Ok, let's use a stack on [ca218a9](https://github.com/GUDHI/gudhi-devel/pull/877/commits/ca218a989b250885e8cd8056f425b566b0df9a64). Order doesn't seem to matter